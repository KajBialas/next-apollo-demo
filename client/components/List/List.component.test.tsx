import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import ListComponent from "./List.component";
jest.mock('@apollo/client');
import * as hooks from './useList.hook';

const mockFetchMore = jest.fn(() => ({}));

const exampleDetailsMock = [
    {
        name: 'Name 1',
        address: 'address 1',
        avatar: 'avatar 1',
        description: 'description 1'

    },
    {
        name: 'Name 2',
        address: 'address 2',
        avatar: 'avatar 2',
        description: 'description 3'
    },
    {
        name: 'Name 3',
        address: 'address 3',
        avatar: 'avatar 3',
        description: 'description 3'

    }
];

describe('List component', () => {
    it('should match snapshot', () => {
        jest.spyOn(hooks, 'useList').mockImplementation(() => ({
            detailsArray: exampleDetailsMock,
            shouldFetchMore: false,
            fetchMoreData: () => {},
            loading: false,
            inputValue: '',
            handleSetInputValue: () => {},
        }));

        const { container } = render(<ListComponent />);
        expect(container).toMatchSnapshot();
    });

    it('should display more button if more results are available', () => {
        jest.spyOn(hooks, 'useList').mockImplementation(() => ({
            detailsArray: exampleDetailsMock,
            shouldFetchMore: true,
            fetchMoreData: () => {},
            loading: false,
            inputValue: '',
            handleSetInputValue: () => {},
        }));

        const { getByTestId } = render(<ListComponent />);
        expect(getByTestId('loadButton')).toBeTruthy();
    });

    it('should not display more button if more results arent available', () => {
        jest.spyOn(hooks, 'useList').mockImplementation(() => ({
            detailsArray: exampleDetailsMock,
            shouldFetchMore: false,
            fetchMoreData: () => {},
            loading: false,
            inputValue: '',
            handleSetInputValue: () => {},
        }));

        const { queryByTestId } = render(<ListComponent />);
        expect(queryByTestId('loadButton')).toBeFalsy();
    });

    it('should call fetchMoreData on button click', () => {
        jest.spyOn(hooks, 'useList').mockImplementation(() => ({
            detailsArray: exampleDetailsMock,
            shouldFetchMore: true,
            fetchMoreData: mockFetchMore,
            loading: false,
            inputValue: '',
            handleSetInputValue: () => {},
        }));

        const { getByTestId } = render(<ListComponent />);
        fireEvent.click(getByTestId('loadButton'));
        expect(mockFetchMore).toHaveBeenCalled();
    });

    it('should show loader if loading is true', () => {
        jest.spyOn(hooks, 'useList').mockImplementation(() => ({
            detailsArray: [],
            shouldFetchMore: false,
            fetchMoreData: mockFetchMore,
            loading: true,
            inputValue: '',
            handleSetInputValue: () => {},
        }));

        const { getByTestId } = render(<ListComponent />);
        expect(getByTestId('loader')).toBeTruthy();
    });

    it('should not show loader if loading is false', () => {
        jest.spyOn(hooks, 'useList').mockImplementation(() => ({
            detailsArray: [],
            shouldFetchMore: false,
            fetchMoreData: mockFetchMore,
            loading: false,
            inputValue: '',
            handleSetInputValue: () => {},
        }));

        const { queryByTestId } = render(<ListComponent />);
        expect(queryByTestId('loader')).toBeFalsy();
    });

    it('should show no data message if records not exists', () => {
        jest.spyOn(hooks, 'useList').mockImplementation(() => ({
            detailsArray: [],
            shouldFetchMore: false,
            fetchMoreData: mockFetchMore,
            loading: false,
            inputValue: '',
            handleSetInputValue: () => {},
        }));

        const { getByTestId } = render(<ListComponent />);
        expect(getByTestId('noDataMessage')).toBeTruthy();
    });

    it('should not show no data message if records exists', () => {
        jest.spyOn(hooks, 'useList').mockImplementation(() => ({
            detailsArray: exampleDetailsMock,
            shouldFetchMore: false,
            fetchMoreData: mockFetchMore,
            loading: false,
            inputValue: '',
            handleSetInputValue: () => {},
        }));

        const { queryByTestId } = render(<ListComponent />);
        expect(queryByTestId('noDataMessage')).toBeFalsy();
    });
});
