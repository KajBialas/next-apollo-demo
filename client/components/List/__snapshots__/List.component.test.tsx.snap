// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should match snapshot 1`] = `
<div>
  <div>
    <div>
      <div
        class="ui input sc-gsTEea fNZppM"
        data-testid="searchBar"
      >
        <input
          name="searchBar"
          placeholder="Search keyword"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="ui grid"
    >
      <div
        class="row"
      >
        <div
          class="four wide computer sixteen wide mobile eight wide tablet column"
          data-testid="cardContainer"
        >
          <div
            class="ui card sc-bdfBQB foBiCW"
          >
            <div
              class="image"
            >
              <img
                src="avatar 1"
              />
            </div>
            <div
              class="content"
            >
              <div
                class="header"
              >
                Name 1
              </div>
              <div
                class="meta"
              >
                Friend
              </div>
              <div
                class="description"
              >
                description 1
              </div>
            </div>
            <div
              class="extra content"
            >
              <div>
                <i
                  aria-hidden="true"
                  class="address book icon"
                />
                address 1
              </div>
            </div>
          </div>
        </div>
        <div
          class="four wide computer sixteen wide mobile eight wide tablet column"
          data-testid="cardContainer"
        >
          <div
            class="ui card sc-bdfBQB foBiCW"
          >
            <div
              class="image"
            >
              <img
                src="avatar 2"
              />
            </div>
            <div
              class="content"
            >
              <div
                class="header"
              >
                Name 2
              </div>
              <div
                class="meta"
              >
                Friend
              </div>
              <div
                class="description"
              >
                description 3
              </div>
            </div>
            <div
              class="extra content"
            >
              <div>
                <i
                  aria-hidden="true"
                  class="address book icon"
                />
                address 2
              </div>
            </div>
          </div>
        </div>
        <div
          class="four wide computer sixteen wide mobile eight wide tablet column"
          data-testid="cardContainer"
        >
          <div
            class="ui card sc-bdfBQB foBiCW"
          >
            <div
              class="image"
            >
              <img
                src="avatar 3"
              />
            </div>
            <div
              class="content"
            >
              <div
                class="header"
              >
                Name 3
              </div>
              <div
                class="meta"
              >
                Friend
              </div>
              <div
                class="description"
              >
                description 3
              </div>
            </div>
            <div
              class="extra content"
            >
              <div>
                <i
                  aria-hidden="true"
                  class="address book icon"
                />
                address 3
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
